### PR TITLE
Repeater - experiment execution across multiple slurm jobs

### PR DESCRIPTION
**Pull request by madragonse**
# Description
Fast implementation 
   - left a few TODO tag comments
   - not tested with non-slurm and repeater combination
   - not tested with other clusters/executors than Writer and Athena
   - not tested with long exps

Neptune:

Here multiple jobs for one exp (https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9cfa0e00-b062-46f9-9357-1c8a17df0932&dash=charts&compare=IwFgDGCcA0zQTLJc4HYRA&compareChartsFilter=loss)
Here same experiment in one job (on Athena) (https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9cfa0e00-b062-46f9-9357-1c8a17df0932&detailsTab=monitoring&shortId=LLMRANDOM-14092&type=run)
In Neptune, under the tag "multiple_jobs_ex" the exps divided into several jobs are shown, the tag "in_one" contains the current version logging everything to one Neptune rune.

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [ ] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [x] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [x] Do I understand the code?
   - [x] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

**Review by kuba-krj — COMMENTED** [review]
Pls add Neptune link with an example grid that was run using the current code. It would be helpful to see:
* two options in grid, each having two consecutive experiments
* example of loading from one of these experiments, again with 2 consecutive runs
just to make sure everything works as expected

File: configs/ms-test/repeater_test.yaml
Comment (crewtool) [inline configs/ms-test/repeater_test.yaml:1]: rename

File: research/conditional/utils/conditional_trainer.py [hunk 1]
```diff
@@ -455,8 +487,13 @@ def _save_weights(self, step):
                 self.save_weights_path,
                 self.rank,
                 step,
+                self.logger,
             )
 
+            return True
+        else:
+            return False
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:495]: I would split decision if we should stop xperitment from saving weights

**Review by crewtool — CHANGES_REQUESTED** [review]
Nicely done!!! 
Left some comments for readibility

File: lizrd/train/load_and_save_model.py [hunk 2]
```diff
@@ -80,12 +115,24 @@ def save_checkpoint(
...
+    try:
+        neptune_logger: Run = [
+            l for l in joint_loggers.loggers if isinstance(l, NeptuneLogger)
+        ][0].instance_logger
+        logger_metadata = {
+            "run_id": neptune_logger._sys_id  # dev TODO do it generally for loggers
+        }
+    except Exception as e:
+        print(f"No Neptune logger, no saving, e: {str(e)}")
+        logger_metadata = None
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py:127]: When NeptuneLogger is not a part of JointLogger

File: research/conditional/utils/conditional_trainer.py [hunk 3]
```diff
@@ -447,6 +455,30 @@ def _save_weights(self, step):
...
+                self.model,
+                self.optimizer,
+                self.scaler,
+                self.save_weights_path,
+                self.rank,
+                step,
+                self.logger,
+            )
+
+    def _repeater_save_weight(
```
[the remark this answers is not in the document]
Comment (madragonse) [reply research/conditional/utils/conditional_trainer.py:469]: _repeater_save_weight dependent on repeater_mode turned on

File: research/conditional/utils/conditional_trainer.py (hunk 1 above)
Comment (madragonse) [reply research/conditional/utils/conditional_trainer.py:495]: I have changed that function name, now stopping experiment propagate from repeater_rerun functionality that additionally saves training and model params.

File: research/conditional/utils/conditional_trainer.py [hunk 4]
```diff
@@ -148,7 +150,8 @@ def train(self, n_steps: int):
             with_modules=True,
         ) as p:
             for step in range(self.start_step, n_steps + 1):
-                self._train_step(step)
+                if self._train_step(step):
+                    break
```
[the remark this answers is not in the document]
Comment (madragonse) [reply research/conditional/utils/conditional_trainer.py:154]: done

File: lizrd/train/load_and_save_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py]: replaced with "if"

File: lizrd/train/load_and_save_model.py [hunk 5]
```diff
@@ -1,20 +1,49 @@
...
+            if not latest_model:
+                print(
+                    f"No model yet saved in ({load_weights_path}), starting new training."
+                )
+                return None
+            load_weights_path = load_weights_path / latest_model.name
+
     print(f"Loading checkpoint from {load_weights_path}...")
-    checkpoint = torch.load(load_weights_path)
+    checkpoint = torch.load(str(load_weights_path))
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py]: done

File: lizrd/grid/grid.py [hunk 6]
```diff
@@ -88,13 +97,27 @@ def create_subprocess_args(
...
+            n_job_repetitions = 1
+            if training_args["repeater_mode"]:
+                total_exp_time = None
+                hours, minutes, seconds = map(int, setup_args["time"].split(":"))
+                total_exp_time = hours * 60 * 60 + minutes * 60 + seconds
+
+                if CLUSTER.max_exp_time < total_exp_time:
+                    n_job_repetitions = ceil(total_exp_time / CLUSTER.max_exp_time)
+
+                    arg_time_to_replace_with = f"{int(CLUSTER.max_exp_time/(60*60))}:{int((CLUSTER.max_exp_time%(60*60))/60)}:{int(CLUSTER.max_exp_time%60)}"
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/grid/grid.py]: done

File: lizrd/grid/grid.py [hunk 7]
```diff
@@ -88,13 +97,27 @@ def create_subprocess_args(
                     (runner_main_function, runner_params)
                 ], interactive_debug_session
 
+            n_job_repetitions = 1
+            if training_args["repeater_mode"]:
+                total_exp_time = None
+                hours, minutes, seconds = map(int, setup_args["time"].split(":"))
+                total_exp_time = hours * 60 * 60 + minutes * 60 + seconds
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/grid/grid.py]: done

File: lizrd/grid/grid.py [hunk 8]
```diff
@@ -88,13 +97,27 @@ def create_subprocess_args(
...
+
+                if CLUSTER.max_exp_time < total_exp_time:
+                    n_job_repetitions = ceil(total_exp_time / CLUSTER.max_exp_time)
+
+                    arg_time_to_replace_with = f"{int(CLUSTER.max_exp_time/(60*60))}:{int((CLUSTER.max_exp_time%(60*60))/60)}:{int(CLUSTER.max_exp_time%60)}"
+                    setup_args["time"] = arg_time_to_replace_with
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/grid/grid.py]: done

File: lizrd/train/load_and_save_model.py [hunk 9]
```diff
@@ -1,20 +1,49 @@
...
 from lizrd.support.misc import generate_random_string
 
 
-def get_checkpoint_from_path(load_weights_path: str) -> str:
+def get_latest_model(dir_path) -> pathlib.Path:
+    dir_path = pathlib.Path(dir_path)
+    all_models_paths = dir_path.glob(r"[0-9]*.pt")
+    latest_model = None
+    latest_step = 0
+    for mp in all_models_paths:
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py:23]: done

File: lizrd/train/load_and_save_model.py [hunk 10]
```diff
@@ -1,20 +1,49 @@
...
+        c_step = int(re.findall(r"\d+", mp.name)[0])
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py:24]: done

File: lizrd/train/load_and_save_model.py [hunk 11]
```diff
@@ -1,20 +1,49 @@
...
     FullStateDictConfig,
     StateDictType,
 )
 
+from lizrd.support.logging import JointLogger, NeptuneLogger
 from lizrd.support.misc import generate_random_string
 
 
-def get_checkpoint_from_path(load_weights_path: str) -> str:
+def get_latest_model(dir_path) -> pathlib.Path:
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py:18]: done

File: lizrd/train/load_and_save_model.py [hunk 12]
```diff
@@ -1,20 +1,49 @@
...
+    dir_path = pathlib.Path(dir_path)
+    all_models_paths = dir_path.glob(r"[0-9]*.pt")
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py:20]: done

File: lizrd/train/load_and_save_model.py [hunk 13]
```diff
@@ -1,20 +1,49 @@
...
+    latest_model = None
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py:21]: done

File: lizrd/train/load_and_save_model.py [hunk 14]
```diff
@@ -50,13 +79,18 @@ def load_scaler_state(
     scaler.load_state_dict(checkpoint["scaler"])
 
 
-def prepare_save_weights_path(path_to_dir: Optional[str]) -> Optional[str]:
+def prepare_save_weights_path(
+    path_to_dir: Optional[str], is_repeater: bool = False
+) -> Optional[str]:
     if path_to_dir is None:
+        if is_repeater:
+            raise Exception("Please specify")
```
[the remark this answers is not in the document]
Comment (madragonse) [reply lizrd/train/load_and_save_model.py:87]: done

File: configs/ms-test/repeater_test.yaml
Comment (madragonse) [reply configs/ms-test/repeater_test.yaml:1]: done

File: lizrd/train/load_and_save_model.py [hunk 15]
```diff
@@ -1,18 +1,47 @@
...
     FullStateDictConfig,
     StateDictType,
 )
 
+from lizrd.support.logging import JointLogger, NeptuneLogger
 from lizrd.support.misc import generate_random_string
 
 
-def get_checkpoint_from_path(load_weights_path: str) -> str:
+def get_latest_checkpoint(dir_path) -> pathlib.Path:
```
Comment (crewtool) [inline lizrd/train/load_and_save_model.py:18]: Please add short comment, that by latest we mean highest step

File: research/conditional/utils/argparse.py [hunk 16]
```diff
@@ -11,6 +11,12 @@ def introduce_parser_arguments(
 ) -> argparse.ArgumentParser:
     # CORE model hyperparameters, almost always specified in baseline configs
     parser.add_argument("--cuda_visible", type=str, default=None)
+    parser.add_argument(
+        "--repeater_mode",
+        action="store_true",
+        help="Used when experiment will last longer than cluster max job time. It repeats jobs for expertiment longer continuation. Combines with periodic model saves.",
+    )
+    # parser.add_argument("--repeater_buffer", type=int, help="In minutes, time before cluster force-clousure that jobs saves and ends itself for next job to continue. Maximu time that model and trainig data needs to be saved, plus MAX time of one step (plus TIME OF VALIDATION that can interfeer!).") #dev TODO, currently fixed 15 min buffer time
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:19]: delete

File: research/conditional/utils/conditional_trainer.py [hunk 17]
```diff
@@ -149,6 +151,12 @@ def train(self, n_steps: int):
         ) as p:
             for step in range(self.start_step, n_steps + 1):
                 self._train_step(step)
+                if (
+                    self._repeater_rerun(step, self.repeater_job_end_time)
+                    if self.repeater_job_end_time
+                    else False
+                ):
+                    break
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:159]: Could you simplify this?

File: research/conditional/utils/conditional_trainer.py [hunk 18]
```diff
@@ -189,6 +197,11 @@ def _train_step(
             self._log_weights_and_gradients(step)
             self._log_auxiliary_losses(aux_info["losses"], step)
         self._save_weights(step)
+        # return (
+        #     self._repeater_rerun(step, self.repeater_job_end_time)
+        #     if self.repeater_job_end_time
+        #     else False
+        # ) #dev
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:204]: remove

File: research/conditional/utils/argparse.py (hunk 16 above)
Comment (madragonse) [reply research/conditional/utils/argparse.py:19]: it is TODO, will be added in the future

File: research/conditional/utils/conditional_trainer.py (hunk 17 above)
Comment (madragonse) [reply research/conditional/utils/conditional_trainer.py]: done

File: research/conditional/utils/conditional_trainer.py (hunk 18 above)
Comment (madragonse) [reply research/conditional/utils/conditional_trainer.py:200]: done

File: research/conditional/utils/conditional_trainer.py [hunk 19]
```diff
@@ -447,6 +451,27 @@ def _save_weights(self, step):
...
+        if repeater_job_end_time and ((repeater_job_end_time - time())) < buffer:
+            if isinstance(self.model, FSDP):
+                # for some reason, setting the model to training mode and
+                # running a forward pass is necessary to be able to save it
+                # in FSDP. God help us. #dev can i delete this, or is it a TODO tag?
+                self.model.train()
+                with torch.no_grad():
+                    _ = self.model(
+                        torch.zeros((self.batch_size, self.cutoff), dtype=torch.int)
+                    )
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:474]: Extract this to separate function `initialize_model_if_fsdp` (or something better)

Comment (madragonse) [reply research/conditional/utils/conditional_trainer.py:474]: added to save_checkpoint fun

**Review by crewtool — APPROVED** [review]
LGTM!
